### PR TITLE
Added methods to __all__ list in __init__.py for some modules

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed

Improved the import structure by adding `__all__` lists to the `__init__.py` files for some classes. This is something that LGTM complains about and one could also see it as a bit of a deficiency that not all modules have this.

#### Other comments
I expect that some tests may fail as I've realized that imports are not always straightforward...

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/core/__init__.py [hunk 1]
```diff
@@ -33,3 +33,54 @@
...
+'expand_power_base', 'expand_power_exp', 'arity',
+
+'evalf', 'PrecisionExhausted', 'N',
+
+'Tuple', 'Dict',
+
+'gcd_terms', 'factor_terms', 'factor_nc',
+
+'evaluate',
+
```
Comment (jksuom) [inline sympy/core/__init__.py]: Lots of blank lines.

Comment (oscargus) [reply sympy/core/__init__.py]: Yes, I kept it in the same style as for `calculus` where each submodule/file has its own section of the list. Easy to change though. (I didn't check any other file, so no idea if it is consistent.)

https://github.com/sympy/sympy/blob/fb98dbb5a50dc13489b365ea6558e1efc5e62377/sympy/calculus/__init__.py#L1-L20

After checking a bit more now, there are not that many module with an `__all__` list... I noticed though that printing use a slightly different approach:

https://github.com/sympy/sympy/blob/master/sympy/printing/__init__.py

**Comment by oscarbenjamin** [comment]
This looks good to me.

I actually think that this does need a release notes entry as some code could break because things are not imported  by default any more (I think that's fine though).

Related issues: #13525 #13592 #5315 #5111 #4674

**Comment by oscargus** [comment]
Thanks for pointing out the related issues! I'll read them and see what I get out of it.

So far, I had to add one additional import (evalf) that probably should have been imported in some other way. That is the only thing I've seen break something. Then, of course, people may import stuff in some unexpected ways, so you are probably correct about the release notes. Should probably try to add the remaining modules as well then so that there is only one release note entry for it.

**Comment by asmeurer** [comment]
You can use `pyflakes` on the files to verify that you didn't miss anything.

**Comment by asmeurer** [comment]
I guess @oscarbenjamin means that submodule names aren't imported any more. For example, currently `from sympy import *` imports `summations`, which is `sympy.concrete.summations`, but after this PR it isn't. I agree it's worth a mention in the release notes. 

Ideally we would have `__all__` defined in every `__init__.py`, including the top-level one, so that no submodules are imported. See the discussions on the issues that @oscarbenjamin linked to.

**Comment by oscargus** [comment]
OK! So should the actual files be imported? Like `summations`? This can be added if required (that is what I did for `evalf` although I'm not sure if it is a good idea or not). But if I understand correctly, no submodules as such should be imported (so not `summations` and not `evalf`).

I've read the discussions and from what I can tell, there is a slight bias towards implementing this although some concerns have been raised.

I think it may make sense to agree on the "correct" format of the `__all__`-part in the files. Either like I've done, with sections divided by empty line, or similar without sections, or as in https://github.com/sympy/sympy/blob/master/sympy/printing/__init__.py

The latter is clearly more tedious to obtain, but at the same time probably easier to read once done.

**Comment by asmeurer** [comment]
No they shouldn't. The modules just pollute the namespace, and cause confusion. 

I fully think we should do this, all the way up to the main `__init__.py`. I think some linting is needed alongside it to make sure that no mistakes are made. I'm not exactly sure how to do that yet, other than by using pyflakes. I think one can compare `dir(module)` to `module.__all__`, but there are difficulties if the module in question is also the name of a file, like with `core` or `logic`.

I don't have any strong opinions on the actual formatting. I think the `sympy/printing/__init__.py` variant looks better personally. It's probably slightly slower to do a bunch of +=, though I doubt by much (but we should test it).

**Comment by oscarbenjamin** [comment]
This looks good. Just needs a release note describing in some way which things might need to be imported differently.

File: sympy/core/__init__.py [hunk 2]
```diff
@@ -33,3 +33,54 @@
...
+'vectorize',
+
+'Lambda', 'WildFunction', 'Derivative', 'diff', 'FunctionClass',
+'Function', 'Subs', 'expand', 'PoleError', 'count_ops',
+'expand_mul', 'expand_log', 'expand_func',
+'expand_trig', 'expand_complex', 'expand_multinomial', 'nfloat',
+'expand_power_base', 'expand_power_exp', 'arity',
+
+'evalf', 'PrecisionExhausted', 'N',
```
Comment (smichr) [inline sympy/core/__init__.py]: ⚠️ IMHO this should *not* be in the general namespace. It is a huge source of confusion that `N(x, n)` is not a synonym for `evalf(x, n)` it is equivalent to `S(x).evalf(n)`. `evalf` takes *binary* precision and requires options to be given. It should really be a private method. Most other functions follow the `foo(x) ~ x.foo()` idiom. `evalf` does not.

Comment (asmeurer) [reply sympy/core/__init__.py]: I didn't even realize there was a *function* `evalf`, distinct from the module or method. Perhaps we should rename it to `_evalf` or `evalf_binary` and treat it as private API. Is it used anywhere outside of evalf.py? I didn't see anywhere, but it's not easy to grep for.

**Comment by asmeurer** [comment]
A small test that only module names are removed. It won't detect for things like `evalf` that are modules replaced with functions, so a better function may be needed.

```py
# write_import_star.py
from subprocess import check_output

def main():
    d = {}
    exec("from sympy import *", d)

    branch = check_output('git rev-parse --abbrev-ref HEAD'.split(), encoding='utf-8').strip()

    filename = f"{branch}-names.txt"
    print(f"Writing to {filename}")

    with open(filename, 'w') as f:
        for name in sorted(d):
            f.write(f"{name}\n")

if __name__ == '__main__':
    main()
```

```py
# test_import_star.py
import sys
from subprocess import check_output

def main():
    if len(sys.argv) != 3:
        sys.exit("Please provide the two names files to compare, like 'test_import_star.py master branch1'")

    file1, file2 = sys.argv[1:]

    with open(file1) as f:
        file1_set = set(f.read().splitlines(keepends=False))

    with open(file2) as f:
        file2_set = set(f.read().splitlines(keepends=False))

    branch = check_output('git rev-parse --abbrev-ref HEAD'.split(), encoding='utf-8').strip()

    def this_branch(file):
        return file == f"{branch}-names.txt"

    d = {}
    exec("from sympy import *", d)

    def print_files(file1, file2, file1_set, file2_set):
        print(f"Names in {file1} but not in {file2}:")
        for name in file1_set - file2_set:
            if this_branch(file1):
                print(name, type(d[name]))
            else:
                print(name)
        print()

    print_files(file1, file2, file1_set, file2_set)
    print_files(file2, file1, file2_set, file1_set)


if __name__ == '__main__':
    main()
```

```bash
$ git checkout AddedAllToInit
$ git merge master # To avoid new names in master since this branch was made
$ python write_import_star.py
Writing to AddedAllToInit-names.txt
$ git checkout master
$ python write_import_star.py
Writing to master-names.txt
$ # Run this on master to have the types printed
$ python test_import_star.py master-names.txt AddedAllToInit-names.txt
Names in master-names.txt but not in AddedAllToInit-names.txt:
containers <class 'module'>
products <class 'module'>
rules <class 'module'>
ordinals <class 'module'>
operations <class 'module'>
inference <class 'module'>
cache <class 'module'>
summations <class 'module'>
coreerrors <class 'module'>
compatibility <class 'module'>
boolalg <class 'module'>
facts <class 'module'>
expr_with_intlimits <class 'module'>
exprtools <class 'module'>
gosper <class 'module'>
expr_with_limits <class 'module'>
relational <class 'module'>
basic <class 'module'>
symbol <class 'module'>
power <class 'module'>
mod <class 'module'>
conditionset <class 'module'>
cse_opts <class 'module'>
multidimensional <class 'module'>
mul <class 'module'>
convolutions <class 'module'>
epathtools <class 'module'>
expr <class 'module'>
cse_main <class 'module'>
decorators <class 'module'>
numbers <class 'module'>
singleton <class 'module'>
traversaltools <class 'module'>
fancysets <class 'module'>
contains <class 'module'>
add <class 'module'>
function <class 'module'>

Names in AddedAllToInit-names.txt but not in master-names.txt:
```

So it looks like all the names that were removed are modules, which is good.

File: sympy/core/__init__.py (hunk 2 above)
Comment (asmeurer) [reply sympy/core/__init__.py]: Apparently `evalf` (from `from sympy import *`) in this branch is still the module. I don't know if that was intentional.

**Comment by asmeurer** [comment]
Simpler version that also checks if any types changed

```py
# write_import_star.py
from subprocess import check_output

def main():
    d = {}
    exec("from sympy import *", d)

    branch = check_output('git rev-parse --abbrev-ref HEAD'.split(), encoding='utf-8').strip()

    filename = f"{branch}-names.txt"
    print(f"Writing to {filename}")

    with open(filename, 'w') as f:
        for name in sorted(d):
            f.write(f"{name}, {type(d[name])}\n")

if __name__ == '__main__':
    main()
```

```py
# test_import_star.py
import sys

def main():
    if len(sys.argv) != 3:
        sys.exit("Please provide the two names files to compare, like 'test_import_star.py master branch1'")

    file1, file2 = sys.argv[1:]

    with open(file1) as f:
        file1_set = set(f.read().splitlines(keepends=False))

    with open(file2) as f:
        file2_set = set(f.read().splitlines(keepends=False))

    def print_files(file1, file2, file1_set, file2_set):
        print(f"Names in {file1} but not in {file2}:")
        for name in file1_set - file2_set:
            print(name)
        print()

    print_files(file1, file2, file1_set, file2_set)
    print_files(file2, file1, file2_set, file1_set)


if __name__ == '__main__':
    main()

```

It's still important to merge this branch with master before testing, to avoid false positives. Here is the result

```
$ python test_import_star.py master-names.txt AddedAllToInit-names.txt
Names in master-names.txt but not in AddedAllToInit-names.txt:
singleton, <class 'module'>
expr_with_intlimits, <class 'module'>
expr, <class 'module'>
epathtools, <class 'module'>
ordinals, <class 'module'>
function, <class 'module'>
boolalg, <class 'module'>
containers, <class 'module'>
symbol, <class 'module'>
multidimensional, <class 'module'>
cache, <class 'module'>
traversaltools, <class 'module'>
inference, <class 'module'>
gosper, <class 'module'>
contains, <class 'module'>
convolutions, <class 'module'>
add, <class 'module'>
expr_with_limits, <class 'module'>
decorators, <class 'module'>
summations, <class 'module'>
cse_main, <class 'module'>
exprtools, <class 'module'>
operations, <class 'module'>
coreerrors, <class 'module'>
mod, <class 'module'>
compatibility, <class 'module'>
mul, <class 'module'>
cse_opts, <class 'module'>
relational, <class 'module'>
products, <class 'module'>
rules, <class 'module'>
fancysets, <class 'module'>
numbers, <class 'module'>
power, <class 'module'>
facts, <class 'module'>
basic, <class 'module'>
conditionset, <class 'module'>

Names in AddedAllToInit-names.txt but not in master-names.txt:

```

So it looks good. `evalf` in particular is not changed at all.

**Comment by asmeurer** [comment]
This was already implemented separately.

**Pull request closed without merging**